app/controllers/hosts_controller.rb

```diff
@@ -451,6 +453,36 @@ def update_multiple_power_state
     redirect_back_or_to hosts_path
   end
 
+  def select_multiple_puppet_proxy
+  end
+
+  def update_multiple_puppet_proxy
+    puppet_proxy_id = params[:puppet][:puppet_proxy_id]
+    proxy = SmartProxy.find_by_id(puppet_proxy_id)
+
+    failed_hosts = []
+
+    @hosts.each do |host|
+      begin
+        host.puppet_proxy = proxy
+        host.save!
+      rescue => error
+        failed_hosts << host.name
+        message = _('Failed to set puppet proxy for %s.') % host
+        Foreman::Logging.exception(message, error)
+      end
+    end
+
+    if failed_hosts.empty?
+      notice _('The puppet proxy of the selected hosts was set to %s') % proxy.name
+    else
+      error n_("The puppet proxy could not be set for host: %{host_names}.",
+               "The puppet proxy could not be set for hosts: %{host_names}.",
+               failed_hosts.count) % {:host_names => failed_hosts.to_sentence}
```

Comment:
This should show the errors that happened IMO - it's not very clear *why* a proxy could've not been applied. Let me illustrate with an example:

Host "timo.goebel" - fully managed, DHCP, DNS. I try to assign a puppet proxy to it through this call. I get an error 'the puppet proxy could not be set for host "timo.goebel". Then I dig on the Foreman logs (I can because I'm an admin!) and find I couldn't change it because the DHCP proxy - that has nothing to do with Puppet - is not available. 

Situations like that could happen so at least the error ought to be shown IMO. You can extract this logic to another method in the controller.